metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -2093,6 +2116,50 @@ Optional<ApiMessageAndVersion> changePartitionReassignment(TopicIdPartition tp,
         return builder.setDefaultDirProvider(clusterDescriber).build();
     }
 
+    /**
+     * Handle legacy configuration alterations.
+     */
+    ControllerResult<Map<ConfigResource, ApiError>> legacyAlterConfigs(
+            Map<ConfigResource, Map<String, String>> newConfigs) {
+        ControllerResult<Map<ConfigResource, ApiError>> result =
+                configurationControl.legacyAlterConfigs(newConfigs, false);
+        return maybeTriggerUncleanLeaderElection(result);
+    }
+
+    /**
+     * Handle incremental configuration alterations.
+     */
+    ControllerResult<Map<ConfigResource, ApiError>> incrementalAlterConfigs(
+            Map<ConfigResource, Map<String, Entry<OpType, String>>> configChanges) {
+        ControllerResult<Map<ConfigResource, ApiError>> result =
+                configurationControl.incrementalAlterConfigs(configChanges, false);
+        return maybeTriggerUncleanLeaderElection(result);
+    }
+
+    /**
+     * If "unclean.leader.election.enable" configurations was enabled dynamically, generating unclean leader election
+     * records for all the partitions in this topic if necessary.
+     */
+    ControllerResult<Map<ConfigResource, ApiError>> maybeTriggerUncleanLeaderElection(
+            ControllerResult<Map<ConfigResource, ApiError>> result) {
+        List<ApiMessageAndVersion> records = new ArrayList<>(result.records());
+        for (ApiMessageAndVersion messageAndVersion : result.records()) {
+            ConfigRecord record = (ConfigRecord) messageAndVersion.message();
+            if (record.name().equals(UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG) && "true".equals(record.value())) {
+                String topicName = record.resourceName();
```

Comment:
This won't work for cases where we're enabling unclean.leader.election for the whole node.

For example, if we enable it for the resource `ConfigResource(type=BROKER, name="")` then it means that all nodes will turn on unclean leader election by default.

Reply:
Good point! Added in https://github.com/apache/kafka/pull/16284/commits/a9df33af3657099fb33c7c97a31d20e12a9600b5 . Also add tests.